### DISCUSSION TITLE
Why `params` and `searchParams` type is different with docs?

### DISCUSSION THREAD

**Discussion by ImBIOS** [vercel/next.js#57463] [App Router] [answered]
https://github.com/vercel/next.js/blob/b27aa579083b6054a578806a839ab555db534309/packages/next/src/build/webpack/plugins/next-types-plugin/index.ts#L162-L166

In the source code, we got it `any`, but in [docs](https://nextjs.org/docs/app/api-reference/file-conventions/page):

```page.tsx
export default function Page({
  params,
  searchParams,
}: {
  params: { slug: string }
  searchParams: { [key: string]: string | string[] | undefined }
}) {
  return <h1>My Page</h1>
}
```

My questions are:

1. Why we use different type here? Is there any benefit from this?
2. Isn't it'd be more efficient if we create Next.js built-in type, and let developer to import and use it directly?
3. I'm thinking of contributing about these things, but I want to know first the reason, so I'm not making useless effort, is this a field where community can contribute?

**Comment by franky47** [comment, marked as answer]
Types in the Next.js repository are generic for all pages, but for a specific page you may define them based on what you actually need.

Here, that Page component lives on a path where there is a `[slug]` directory, so the params can be typed as such to access it. 

Same goes for searchParams, but here you don't know at build time what you may receive from the query string at runtime, so the record uses a generic key:value type. You can specify known queries if your app requires them to help a bit with type safety:

```ts
export default function Page({
  searchParams,
}: {
  searchParams: { 
   query: string | string[] | undefined
   page: string | string[] | undefined
   // etc..
  }
}) {
  return <h1>My Page</h1>
}
```

[Typed routes](https://nextjs.org/docs/app/api-reference/next-config-js/typedRoutes) might help automate this, but they are still an experimental feature.
